sdk/dotnet/Utils.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Copyright 2016-2019, Pulumi Corporation.
+
+using System;
+using System.Text;
+
+namespace Pulumi.Docker
+{
+    internal static class Utils
+    {
+        public static (string, string?) GetImageNameAndTag(string baseImageName)
+        {
+            // From https://docs.docker.com/engine/reference/commandline/tag
+            //
+            // "A tag name must be valid ASCII and may contain lowercase and uppercase letters, digits,
+            // underscores, periods and dashes. A tag name may not start with a period or a dash and may
+            // contain a maximum of 128 characters."
+            //
+            // So it is safe for us to just look for the colon, and consume whatever follows as the tag
+            // for the image.
+
+            var lastColon = baseImageName.LastIndexOf(":");
+            var imageName = lastColon < 0 ? baseImageName : baseImageName.Substring(0, lastColon);
+            var tag = lastColon < 0 ? null : baseImageName.Substring(lastColon + 1);
+
+            return (imageName, tag);
+        }
+
+        /// <summary>
+        /// Convert an argument array to an argument string for using with Process.StartInfo.Arguments.
+        /// </summary>
+        public static string EscapeArguments(params string[] args)
+        {
+            var argEnumerator = args.GetEnumerator();
+            var arguments = new StringBuilder();
+
+            if (!argEnumerator.MoveNext())
+            {
+                return string.Empty;
+            }
+
+            arguments.Append(EscapeArguments((string)argEnumerator.Current!));
+
+            while (argEnumerator.MoveNext())
+            {
+                arguments.Append(' ');
+                arguments.Append(EscapeArguments((string)argEnumerator.Current!));
+            }
+
+            return arguments.ToString();
+        }
+
+        /// <summary>
+        /// Convert an argument array to an argument string for using with Process.StartInfo.Arguments.
+        /// </summary>
+        private static string EscapeArguments(string argument)
+        {
+            using var characterEnumerator = argument.GetEnumerator();
```

Comment:
i'm not getting the need for manual enumeratoin.  You only seem to process a single char at a time.  so just do a `foreach (var c in argument) ...`

Reply:
Changed